src/doc/trpl/dining-philosophers.md

```diff
@@ -0,0 +1,942 @@
+problem:
+
+1. A philosopher picks up the fork on their left.
+2. They then pick up the fork on their right.
+3. They eat.
+4. They return the forks.
+
+Now, let’s imagine this sequence of events:
+
+1. Philosopher 1 begins the algorithm, picking up the fork on their left.
+2. Philosopher 2 begins the algorithm, picking up the fork on their left.
+3. Philosopher 3 begins the algorithm, picking up the fork on their left.
+4. Philosopher 4 begins the algorithm, picking up the fork on their left.
+5. Philosopher 5 begins the algorithm, picking up the fork on their left.
+6. ... ? All the forks are taken, but nobody can eat!
+
+There are different ways to solve this problem. We’ll get to our solution in
+the tutorial itself. For now, let’s get started modelling the problem itself.
+We’ll start with the philosophers:
+
+```rust
+struct Philosopher {
+    name: String,
+}
+
+impl Philosopher {
+    fn new(name: &str) -> Philosopher {
+        Philosopher {
+            name: name.to_string(),
+        }
+    }
+}
+
+fn main() {
+    let p1 = Philosopher::new("Baruch Spinoza");
+    let p2 = Philosopher::new("Gilles Deleuze");
+    let p3 = Philosopher::new("Karl Marx");
+    let p4 = Philosopher::new("Friedrich Nietzsche");
+    let p5 = Philosopher::new("Michel Foucault");
+}
+```
+
+Here, we make a [`struct`][struct] to represent a philosopher. For now,
+a name is all we need. We choose the [`String`][string] type for the name,
+rather than `&str`. Generally speaking, working with a type which owns its
+data is easier than working with one that uses references. If we wanted to use
+`&str`s in our `struct`, we’d have to introduce some extra notation:
+
+```rust
+struct Philosopher<'a> {
+    name: &'a str,
+}
+```
+
+[struct]: structs.html
+[string]: strings.html
+
+These `'a` markers are called [‘lifetimes’][lifetimes], and help Rust understand
+the scope in which a lifetime is valid. You’ll notice the annotation in two places:
+
+```rust
+struct Philosopher<'a> {
+#    name: &'a str,
+# }
+```
+
+[lifetimes]: lifetimes.html
+
+This part says “The `Philosopher` struct has a lifetime parameter named `’a`,”
+and
+
+```rust
+# struct Philosopher<'a> {
+name: &'a str,
+# }
+```
+
+This part says “The `name` field is a string reference with the lifetime `’a`.
+
+With a small `struct` like this, it seems kind of obvious: there’s only one
+reference. But consider these two structs:
+
+```rust
+struct Philosopher1<'a, 'b> {
+    first_name: &'a str,
+    last_name: &'b str,
+}
+
+struct Philosopher2<'a> {
+    first_name: &'a str,
+    last_name: &'a str,
+}
+```
+
+The first version says that we have two references that are valid for different
+scopes, but the second says they’re valid for the same scope.
+
+In any case, we’re getting off track. We don’t want references, we want the
+struct to ‘have ownership’ over the value. The `String` type gives us this:
+when our `Philosopher` goes out of scope, we’ll free the name as well.
```

Comment:
The explanation that we could've used `&'a str` instead of `String` is important if you want to learn Rust, but also too long to be then called 'off track', IMHO. If it's not important for the current guide (you even explain some `&str` later on), why not make it a (relatively long) footnote instead? This way it doesn't interfere with the progress of the guide unless you actually want to read it.